File: packages/checks/src/checks/noBlankLinesBetweenSingleLineDeclarations.ts
import { Array, Function, HashSet, Option, Struct, pipe } from "effect"
import * as ts from "typescript"
import { isDeclarationStatement, isStatementContainer } from "./support/tsNode.js"
import type { CheckContext } from "@better-typescript/core/engine/check/data"
import type { Check } from "@better-typescript/core/engine/check/data"
import type { Detection } from "@better-typescript/core/engine/location/data"
import type { NonEmptyRefactorExamples } from "@better-typescript/core/engine/example/data"

import { fixtureRefactorExamples } from "../fixtureExamples.js"
import { nodeCheck, detection } from "@better-typescript/core/engine/check"

const singleLineDeclarationKindList: ReadonlyArray<ts.SyntaxKind> = Array.make(
  ts.SyntaxKind.VariableStatement,
  ts.SyntaxKind.FunctionDeclaration,
  ts.SyntaxKind.ClassDeclaration,
  ts.SyntaxKind.InterfaceDeclaration,
  ts.SyntaxKind.TypeAliasDeclaration,
  ts.SyntaxKind.EnumDeclaration,
  ts.SyntaxKind.ModuleDeclaration
)

const functionLikeKinds = HashSet.make(
  ts.SyntaxKind.FunctionDeclaration,
  ts.SyntaxKind.FunctionExpression,
  ts.SyntaxKind.ArrowFunction,
  ts.SyntaxKind.MethodDeclaration,
  ts.SyntaxKind.Constructor,
  ts.SyntaxKind.GetAccessor,
  ts.SyntaxKind.SetAccessor
)

const message = "Single-line declarations must not have blank lines between them."

const hint =
  "Remove the empty line between these adjacent single-line declarations so they " +
  "stay contiguous. Blank lines remain required around multi-line declarations; " +
  "keep those separators when a neighbor is multi-line."

const singleLineBlankLinePattern = /\n[ \t]*\r?\n/

const fallbackFalse = Function.constant(false)

const fallbackMissingIndex = Function.constant(-1)

const isFunctionLike = (node: ts.Node): boolean => HashSet.has(functionLikeKinds, node.kind)

const parentContinuesFunctionSearch = (parent: ts.Node): boolean => {
  const parentIsFunction = isFunctionLike(parent)
  const parentIsSourceFile = ts.isSourceFile(parent)
  const parentIsNotFunction = parentIsFunction === false
  const parentIsNotSourceFile = parentIsSourceFile === false
  const continueSearchConditions = Array.make(parentIsNotFunction, parentIsNotSourceFile)

  return Array.every(continueSearchConditions, Boolean)
}

const isInsideFunction = (node: ts.Node): boolean =>
  pipe(
    Option.fromNullishOr(node.parent),
    Option.map((parent) => {
      const parentIsFunction = isFunctionLike(parent)
      const continueSearch = parentContinuesFunctionSearch(parent)
      const nestedHit = continueSearch && isInsideFunction(parent)
      const hitConditions = Array.make(parentIsFunction, nestedHit)

      return Array.some(hitConditions, Boolean)
    }),
    Option.getOrElse(fallbackFalse)
  )

const contiguousSingleLineBlankLineMatches = (context: CheckContext) => {
  const match = detection(context)
  const sourceFile = context.sourceFile
  const text = sourceFile.getFullText()

  const matches = (node: ts.Node): ReadonlyArray<Detection> => {
    if (!isDeclarationStatement(node)) {
      return Array.empty()
    }

    const startPosition = node.getStart(sourceFile)
    const endPosition = node.getEnd()
    const start = sourceFile.getLineAndCharacterOfPosition(startPosition)
    const end = sourceFile.getLineAndCharacterOfPosition(endPosition)
    const currentIsSingleLine = end.line === start.line
    const insideFunction = isInsideFunction(node)
    const parent = node.parent

    const siblingsOption = pipe(
      Option.liftPredicate(isStatementContainer)(parent),
      Option.map(Struct.get("statements"))
    )

    const hasBlankLineAfterPreviousSingleLine = pipe(
      siblingsOption,
      Option.map((siblings) => {
        const index = pipe(
          Array.findFirstIndex(siblings, (sibling) => sibling === node),
          Option.getOrElse(fallbackMissingIndex)
        )

        const previous = Array.get(siblings, index - 1)

        return pipe(
          previous,
          Option.map((prev) => {
            const previousIsDeclaration = isDeclarationStatement(prev)
            const previousStartPosition = prev.getStart(sourceFile)
            const previousEndPosition = prev.getEnd()
            const previousStart = sourceFile.getLineAndCharacterOfPosition(previousStartPosition)
            const previousEnd = sourceFile.getLineAndCharacterOfPosition(previousEndPosition)
            const previousIsSingleLine = previousEnd.line === previousStart.line
            const beforeEnd = prev.getEnd()
            const afterStart = node.getStart(sourceFile)
            const between = text.slice(beforeEnd, afterStart)
            const hasBlankLine = singleLineBlankLinePattern.test(between)

            const gapConditions = Array.make(
              previousIsDeclaration,
              previousIsSingleLine,
              hasBlankLine
            )

            return Array.every(gapConditions, Boolean)
          }),
          Option.getOrElse(fallbackFalse)
        )
      }),
      Option.getOrElse(fallbackFalse)
    )

    const flagConditions = Array.make(
      insideFunction,
      currentIsSingleLine,
      hasBlankLineAfterPreviousSingleLine
    )

    const shouldFlag = Array.every(flagConditions, Boolean)

    const blankLineMatch = match({
      node,
      message,
      hint
    })

    return shouldFlag ? Array.of(blankLineMatch) : Array.empty()
  }

  return matches
}

const check = nodeCheck(singleLineDeclarationKindList)(isDeclarationStatement)(
  contiguousSingleLineBlankLineMatches
)

export const noBlankLinesBetweenSingleLineDeclarations: Check = check

export const noBlankLinesBetweenSingleLineDeclarationsExamples: NonEmptyRefactorExamples =
  fixtureRefactorExamples("no-blank-lines-between-single-line-declarations")

File: internal/rules/no_blank_lines_between_single_line_declarations/no_blank_lines_between_single_line_declarations.go
package no_blank_lines_between_single_line_declarations

import (
	"regexp"
	"strings"

	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/andrueandersoncs/better-typescript/internal/utils"
	"github.com/microsoft/typescript-go/shim/ast"
)

var blankLinePattern = regexp.MustCompile(`\n[ \t]*\r?\n`)

var message = rule.RuleMessage{
	Id:          "noBlankLinesBetweenSingleLineDeclarations",
	Description: "Single-line declarations must not have blank lines between them.",
	Help:        "Remove the empty line between these adjacent single-line declarations so they stay contiguous. Blank lines remain required around multi-line declarations; keep those separators when a neighbor is multi-line.",
}

func isDeclarationStatement(node *ast.Node) bool {
	switch node.Kind {
	case ast.KindVariableStatement, ast.KindFunctionDeclaration, ast.KindClassDeclaration, ast.KindInterfaceDeclaration, ast.KindTypeAliasDeclaration, ast.KindEnumDeclaration, ast.KindModuleDeclaration:
		return true
	default:
		return false
	}
}

func isInsideFunction(node *ast.Node) bool {
	for parent := node.Parent; parent != nil && !ast.IsSourceFile(parent); parent = parent.Parent {
		if ast.IsFunctionLike(parent) {
			return true
		}
	}
	return false
}

func isSingleLine(sourceFile *ast.SourceFile, node *ast.Node) bool {
	r := utils.TrimNodeTextRange(sourceFile, node)
	source := sourceFile.Text()
	return !strings.Contains(source[r.Pos():r.End()], "\n")
}

func checkDeclaration(ctx rule.RuleContext, node *ast.Node) {
	if !isDeclarationStatement(node) || !isInsideFunction(node) || !isSingleLine(ctx.SourceFile, node) {
		return
	}
	parent := node.Parent
	if parent == nil || !parent.CanHaveStatements() {
		return
	}
	statements := parent.Statements()
	for index, current := range statements {
		if current != node || index == 0 {
			continue
		}
		previous := statements[index-1]
		if !isDeclarationStatement(previous) || !isSingleLine(ctx.SourceFile, previous) {
			return
		}
		beforeEnd := previous.End()
		afterStart := utils.TrimNodeTextRange(ctx.SourceFile, node).Pos()
		if beforeEnd <= afterStart && blankLinePattern.MatchString(ctx.SourceFile.Text()[beforeEnd:afterStart]) {
			ctx.ReportNode(node, message)
		}
		return
	}
}

func listener(ctx rule.RuleContext) func(*ast.Node) {
	return func(node *ast.Node) { checkDeclaration(ctx, node) }
}

var NoBlankLinesBetweenSingleLineDeclarationsRule = rule.Rule{
	Name: "no-blank-lines-between-single-line-declarations",
	Run: func(ctx rule.RuleContext, options any) rule.RuleListeners {
		check := listener(ctx)
		return rule.RuleListeners{
			ast.KindVariableStatement:    check,
			ast.KindFunctionDeclaration:  check,
			ast.KindClassDeclaration:     check,
			ast.KindInterfaceDeclaration: check,
			ast.KindTypeAliasDeclaration: check,
			ast.KindEnumDeclaration:      check,
			ast.KindModuleDeclaration:    check,
		}
	},
}
